src/database.ts

```diff
@@ -1,389 +1,383 @@
-			return runAction({
-				schema,
-				store,
-				mode: 'readwrite',
-				routine: (_store) =>
-					usingDatabase(store, () => handleRequest(_store.delete(id), item)
-							.then(() => true)
-							.catch(() => false))
-			});
-		}
-		const localItem: T & { flag?: string } = {...item, flag: 'D'};
-		return runAction<boolean>({
-			store,
-			schema,
-			mode: 'readwrite',
-			routine: (_store) => handleRequest(_store.put(localItem), true)
-				.then(((res) => res === undefined ? true : res))
-		});
-	}
-
-	private async deleteRemote(id: IndexedKey, item: T): Promise<boolean> {
-		if (!this.api) {
-			return false;
-		}
-		const isOnline = isOnlineSupport(this.api) ? await this.api.isOnline() : undefined;
-		if (isOnline === undefined || isOnline) {
-			return this.api.delete(item).then(async (deleted: boolean) => {
-				return deleted && await this.deleteLocal(id, item, false);
-			}).catch((err) => {
-				console.error(`can't call delete of api ${buildEntityIdent(this.schema)}`, err);
-				return false;
-			});
-		}
-		return true;
-	}
-
-	private getId<S>(item: T, callback: (id: IndexedKey) => Promise<S>): Promise<S> {
-		const {keyPath = 'id'} = this.schema;
-		if (!(item as any).hasOwnProperty(keyPath)) {
-			return Promise.reject();
-		}
-		const id: IndexedKey = getValue(item, keyPath as keyof T) as any;
-		return callback(id);
-	}
+  constructor(
+    private objectStore: IDBObjectStore,
+    private key: string,
+    private schema: StoreSchema,
+    private api?: CrudApi<T>
+  ) {}
+
+  /**
+   * push local transactions to remote and try to fetch updates from remote
+   */
+  public sync(): Promise<void> {
+    type F = WithFlag<T>;
+    return new Promise(async (done, _) => {
+      const isOnline = this.api && isOnlineSupport(this.api)
+        ? await this.api.isOnline()
+        : undefined;
+      if (isOnline === false) {
+        done();
+        return;
+      }
+      const localItems = new Promise<F[]>((resolve, reject) => {
+        this.prepareTransaction();
+        const req = this.objectStore.index('flag').getAll();
+        req.onsuccess = () => resolve(req.result as F[]);
+        req.onerror = () => reject();
+      });
+      const actions = (await localItems).map(
+        (item: F): Promise<unknown> => {
+          switch (item.flag) {
+            case 'D':
+              return this.getId(item, (id: T[keyof T]) =>
+                this.deleteRemote(id, item)
+              );
+            case 'C':
+              return this.createRemote(item);
+            case 'U':
+              return this.updateRemote(item);
+            default:
+              return Promise.resolve();
+          }
+        }
+      );
+      await Promise.all(actions);
+      if (this.api && (isOnline === undefined || isOnline)) {
+        await this.api
+          .getAll()
+          .then(fromRemote => this.updateLocalStorage(fromRemote))
+          .catch((err) => console.error(`${this.key} sync failed with api`, err));
```

Comment:
```suggestion
        try {
          const fromRemote = await this.api.getAll();
          this.updateLocalStorage(fromRemote);
        } catch(error) {
          console.error(`${this.key} sync failed with api`, error);
        }
```

Please always use async await to make sure to have good async stack traces.

I also suggest to not only log the error but to actually reject in error cases.